core/src/main/scala/org/apache/spark/util/SizeEstimator.scala

```diff
@@ -316,62 +338,40 @@ object SizeEstimator extends Logging {
-    val sizeCount = Array.fill(fieldSizes.max + 1)(0)
+    var fieldOffsets = parent.fieldOffsets.toList
+
+    var index = 0
 
-    // iterate through the fields of this class and gather information.
-    for (field <- cls.getDeclaredFields) {
+    while (index < fieldCount) {
+      val field = fields(index)
       if (!Modifier.isStatic(field.getModifiers)) {
         val fieldClass = field.getType
         if (fieldClass.isPrimitive) {
-          sizeCount(primitiveSize(fieldClass)) += 1
+          if (cls == classOf[Double] || cls == classOf[Long]) {
+            shellSize += 8
+          } else {
+            shellSize += 4
+          }
         } else {
-          field.setAccessible(true) // Enable future get()'s on this field
-          sizeCount(pointerSize) += 1
-          pointerFields = field :: pointerFields
+          shellSize += pointerSize
+          fieldOffsets = Unsafe.instance.objectFieldOffset(field).toInt :: fieldOffsets
         }
       }
+      index += 1
     }
 
-    // Based on the simulated field layout code in Aleksey Shipilev's report:
-    // http://cr.openjdk.java.net/~shade/papers/2013-shipilev-fieldlayout-latest.pdf
-    // The code is in Figure 9.
-    // The simplified idea of field layout consists of 4 parts (see more details in the report):
-    //
-    // 1. field alignment: HotSpot lays out the fields aligned by their size.
-    // 2. object alignment: HotSpot rounds instance size up to 8 bytes
-    // 3. consistent fields layouts throughout the hierarchy: This means we should layout
-    // superclass first. And we can use superclass's shellSize as a starting point to layout the
-    // other fields in this class.
-    // 4. class alignment: HotSpot rounds field blocks up to to HeapOopSize not 4 bytes, confirmed
-    // with Aleksey. see https://bugs.openjdk.java.net/browse/CODETOOLS-7901322
-    //
-    // The real world field layout is much more complicated. There are three kinds of fields
-    // order in Java 8. And we don't consider the @contended annotation introduced by Java 8.
-    // see the HotSpot classloader code, layout_fields method for more details.
-    // hg.openjdk.java.net/jdk8/jdk8/hotspot/file/tip/src/share/vm/classfile/classFileParser.cpp
-    var alignedSize = shellSize
-    for (size <- fieldSizes if sizeCount(size) > 0) {
-      val count = sizeCount(size).toLong
-      // If there are internal gaps, smaller field can fit in.
-      alignedSize = math.max(alignedSize, alignSizeUp(shellSize, size) + size * count)
-      shellSize += size * count
-    }
-
-    // Should choose a larger size to be new shellSize and clearly alignedSize >= shellSize, and
-    // round up the instance filed blocks
-    shellSize = alignSizeUp(alignedSize, pointerSize)
-
     // Create and cache a new ClassInfo
-    val newInfo = new ClassInfo(shellSize, pointerFields)
-    classInfos.put(cls, newInfo)
+    val newInfo = new ClassInfo(shellSize, fieldOffsets.toArray)
```

Comment:
We are loosing out on padding due to allignment here which the earlier code was computing. No ?